tests/WE2E/utils.py

```diff
@@ -0,0 +1,584 @@
+    for key in links.keys():
+        link = links[key]
+        testdict[link[2]]["alternate_name"] = link[0]
+        testdict[link[2]]["alternate_directory_name"] = link[1]
+
+    # Print the file
+    with open(txtfile, 'w', encoding="utf-8") as f:
+        # Field delimiter character
+        d = "\" | \""
+        txt_output = ['"Test Name']
+        txt_output.append(f'(Subdirectory){d}Alternate Test Names')
+        txt_output.append(f'(Subdirectories){d}Test Purpose/Description{d}Relative Cost of Running Dynamics')
+        txt_output.append(f'(1 corresponds to running a 6-hour forecast on the RRFS_CONUS_25km predefined grid using the default time step){d}PREDEF_GRID_NAME{d}CCPP_PHYS_SUITE{d}EXTRN_MDL_NAME_ICS{d}EXTRN_MDL_NAME_LBCS{d}DATE_FIRST_CYCL{d}DATE_LAST_CYCL{d}INCR_CYCL_FREQ{d}FCST_LEN_HRS{d}DT_ATMOS{d}LBC_SPEC_INTVL_HRS{d}NUM_ENS_MEMBERS')
+
+        for line in txt_output:
+            f.write(f"{line}\n")
+        for expt in testdict:
+            f.write(f"\"{expt}\n(")
+            f.write(f"{testdict[expt]['directory']}){d}")
+            if "alternate_name" in testdict[expt]:
+                f.write(f"{testdict[expt]['alternate_name']}\n"\
+                        f"({testdict[expt]['alternate_directory_name']}){d}")
+            else:
+                f.write(f"{d}\n")
+            desc = testdict[expt]['metadata']['description'].splitlines()
+            for line in desc[:-1]:
+                f.write(f"    {line}\n")
+            f.write(f"    {desc[-1]}")
+            #Write test relative cost and number of test forecasts (for cycling runs)
+            f.write(f"{d}'{round(testdict[expt]['cost'],2)}{d}'{round(testdict[expt]['num_fcsts'])}")
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'workflow','PREDEF_GRID_NAME'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'workflow','CCPP_PHYS_SUITE'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'task_get_extrn_ics',
+                                                'EXTRN_MDL_NAME_ICS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'task_get_extrn_lbcs',
+                                                'EXTRN_MDL_NAME_LBCS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'workflow','DATE_FIRST_CYCL'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'workflow','DATE_LAST_CYCL'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'workflow','INCR_CYCL_FREQ'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'workflow','FCST_LEN_HRS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'task_run_fcst','DT_ATMOS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'task_get_extrn_lbcs',
+                                                'LBC_SPEC_INTVL_HRS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'global','NUM_ENS_MEMBERS') + "\n")
```

Comment:
This bit is pretty clunky. Just to avoid repetition and inefficient calls to a function that does what a built-in does, you might consider:

```
key_pairs = [
 ('workflow', 'PREDEF_GRID_NAME'),
 ('workflow', 'CCPP_PHYS_SUITE'),
  ...
  ]

for key1, key2 in key_pairs:
    f.write(f"{d} {testdict[expt].get(key1, {}).get(key2, '')}"
```